src/CredentialSource/AwsNativeSource.php

```diff
@@ -0,0 +1,314 @@
+<?php
+/*
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Auth\CredentialSource;
+
+use Google\Auth\ExternalAccountCredentialSourceInterface;
+use Google\Auth\HttpHandler\HttpClientCache;
+use Google\Auth\HttpHandler\HttpHandlerFactory;
+use GuzzleHttp\Psr7\Request;
+
+/**
+ * Authenticates requests using IAM credentials.
+ */
+class AwsNativeSource implements ExternalAccountCredentialSourceInterface
+{
+    private const CRED_VERIFICATION_QUERY = 'Action=GetCallerIdentity&Version=2011-06-15';
+
+    private string $audience;
+    private string $regionUrl;
+    private string $regionalCredVerificationUrl;
+    private ?string $securityCredentialsUrl;
+
+    public function __construct(
+        string $audience,
+        string $regionUrl,
+        string $regionalCredVerificationUrl,
+        string $securityCredentialsUrl = null
+    ) {
+        $this->audience = $audience;
+        $this->regionUrl = $regionUrl;
+        $this->regionalCredVerificationUrl = $regionalCredVerificationUrl;
+        $this->securityCredentialsUrl = $securityCredentialsUrl;
+    }
+
+    public function fetchSubjectToken(callable $httpHandler = null): string
+    {
+        if (is_null($httpHandler)) {
+            $httpHandler = HttpHandlerFactory::build(HttpClientCache::getHttpClient());
+        }
+
```

Comment:
Should support Imsdv2 session token: https://github.com/googleapis/google-auth-library-nodejs/pull/1363, https://github.com/googleapis/google-auth-library-java/pull/850 this is used to get another header value that needs to be used for the region and security credentials call

Reply:
I've added support for this, but using the credentials that I received for testing (which do not have the `imdsv2_session_token_url` property in the JSON), this is now failing because the library fails to fetch the region from the server without the URL. 

This would imply that the parameter is required in cases where the signing vars are not set as environment variables. If this is the case, I think my app should throw an exception saying "imdsv2_session_token_url not found in credentials and is required when signing vars are not present as environment variables", or something along these lines. WDYT?

Reply:
The parameter is not always required when the url is used, since users do not need to have imsdv2 enforced on their EC2 instances, so we shouldn't throw that error

Reply:
After speaking with @aeitzman, the answer is that this would be an edgecase error due to a user mistake (such as my own) due to the wrong configuration being used on a machine it isn't meant for. We can handle this error if we'd like, but I agree that it's probably best to just throw the exception as-is.